plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -723,21 +724,17 @@ enterComputeCek = computeCek (toWordArray 0) where
         computeCek unbudgetedSteps ctx (Env.cons arg env) body
     -- Annotating @f@ and @exF@ with bangs gave us some speed-up, but only until we added a bang to
     -- 'VCon'. After that the bangs here were making things a tiny bit slower and so we removed them.
-    applyEvaluate !unbudgetedSteps !ctx (VBuiltin fun term (BuiltinRuntime sch f exF)) arg = do
+    applyEvaluate !unbudgetedSteps !ctx (VBuiltin fun term runtime) arg = do
         let argTerm = dischargeCekValue arg
```

Comment:
It's used lazily, and only in constructing errors. It's always been like that.

Reply:
> If we pass a term argument as a branch of ifThenElse for example, is it the case that it can contain variables referring to things in environments when it goes in, but it'll have been (lazily) discharged when it comes out of the other end so that the variables will be gone if we look at the term?

No, a `CekValue` going to `IfThenElse` comes out unaffected. We only have this discharged `Term` for recreating the Plutus version of the partial application **in case we never saturate the application** and still use it either as a part of the result or in an error message.

I.e. this line is the whole point of that `Term` in `VBuiltin`:

```haskell
dischargeCekValue :: CekValue uni fun -> Term NamedDeBruijn uni fun ()
dischargeCekValue = \case
    <...>
    VBuiltin _ term _                    -> term
```

If a builtin application ever gets saturated, the `Term` part of `VBuiltin` gets thrown away without ever being forced (hence why it's so important to keep it lazy), since we don't need a Plutus term that has computed away.

Reply:
> It's used lazily, and only in constructing errors.

And if the result of evaluation has partial applications inside of it.

> It's always been like that.

Kenneth correctly pinpointed the exact commit. We started doing that 15 months ago when we switched from the original builtin application machinery (collects arguments in a list, feed them all at once to the builtin) to the alternative one (feed arguments one by one directly to the builtin without storing them in any intermediate data structure). With lists-of-arguments it was trivial to reconstruct the partial application, hence we didn't have that lazy field in `VBuiltin`.

Reply:
>  We only have this discharged Term for recreating the Plutus version of the partial application in case we never saturate the application and still use it either as a part of the result or in an error message.

[Me]
>  Or have I just got the wrong end of the stick altogether?

It appears that I had.  I'd managed to forget that `BuiltinRuntime` was collecting the arguments itself, so I just got completely confused.  Thanks for the clarification (and sorry for the obtuseness)!